File: 985-bag-of-tokens/bag-of-tokens.java
class Solution {
    public int bagOfTokensScore(int[] tokens, int power) {
        Arrays.sort(tokens);
        int n=tokens.length,i=0,j=n-1;
        int ans=0,score=0;
        while(i<=j){
            if(tokens[i]<=power){
                score++;
                power-=tokens[i];
                i++;
            }
            else if(score>=1){
                score--;
                power+=tokens[j];
                j--;
            }
            else{
                break;
            }
            ans=Math.max(ans,score);
        }
        return ans;
    }
}

File: 985-bag-of-tokens/bag-of-tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(),tokens.end());
        int i=0,j=tokens.size()-1;
        int score=0;
        int ans=0;
        while(i<=j){
            if(power>=tokens[i]){
                score++;
                power-=tokens[i];
                ans=max(ans,score);
                i++;
            }
            else if(score){
                score--;
                power+=tokens[j];
                j--;
            }
            else{
                break;
            }
        }
        return ans;
    }
};